service/ca/pki.go

```diff
@@ -3,14 +3,39 @@ package ca
 import (
 	"fmt"
 	"strings"
+	"time"
 
+	"github.com/cenkalti/backoff"
 	"github.com/giantswarm/certctl/service/pki"
 	"github.com/giantswarm/certctl/service/token"
 	"github.com/giantswarm/certificatetpr"
 	microerror "github.com/giantswarm/microkit/error"
 )
 
-// SetupPKI creates a PKI backend and policy if one does not exists for the cluster.
+const (
+	setupPKIMaxElapsedTime = 30 * time.Second
+)
+
+// SetupPKIAndWait creates a PKI backend and policy if one does not exist for
+// the cluster. If an error occurs an exponential backoff is used. After the
+// max elapsed time the error is returned to the caller.
+func (s *Service) SetupPKIAndWait(cert certificatetpr.Spec) error {
+	initBackoff := backoff.NewExponentialBackOff()
+	initBackoff.MaxElapsedTime = setupPKIMaxElapsedTime
+
+	operation := func() error {
+		err := s.SetupPKI(cert)
+		if err != nil {
+			s.Logger.Log("info", "failed to setup PKI - retrying")
```

Comment:
My thinking was that SetupPKI would have the normal logging. In the wait wrapper I just wanted to indicate that a retry was taking place.

But it is different to our normal error handling and I get the point about returning the error. I'll take another look.